Adding expose application

maayarosama opened the pull request:
### Description

Adding Domains solution to the dashboard
![Screenshot from 2024-08-18 15-15-19](https://github.com/user-attachments/assets/4d3255fc-9473-40bf-8378-df48f38ef7f6)
![Screenshot from 2024-08-18 15-27-42](https://github.com/user-attachments/assets/f5cf87b5-720f-4631-b602-47289981d089)


### Related Issues

- #3178
### Documentation PR

- https://github.com/threefoldtech/info_grid/pull/610
- https://github.com/threefoldtech/info_grid/pull/627
### Checklist

- [ ] Tests included
- [x] Build pass
- [x] Documentation
- [x] Code format and docstrings
- [x] Screenshots/Video attached (needed for UI changes)

maayarosama committed:
Adding expose application

maayarosama committed:
Removing selection details output

maayarosama committed:
Adding validations to ip field and expose icon

mik-tf committed:
adjused expose descriptions

mik-tf commented:
Note: I adjusted 2 basic texts on expose. To say: Expose allows... instead of Expose application allows... 

Just like we say, e.g.,  Gitea provides... instead of Gitea application provides...

Commit: https://github.com/threefoldtech/tfgrid-sdk-ts/pull/3215/commits/46b4c5371920ae9a61dbc5b8f087087bab797db9

mik-tf commented:
@maayarosama Sorry for the direct commit. I understand that I should have made a review/proposition. Will use reviews in the future! Great work btw :)

amiraabouhadid reviewed (commented):
please add a validation that ip should not be local host 
![image](https://github.com/user-attachments/assets/f81335fd-a968-40a3-9912-2f8b3181a7b7)

AhmedHanafy725 commented on packages/playground/src/weblets/tf_expose.vue:
```diff
@@ -0,0 +1,242 @@
+        v-model="solution"
+        :small="{ cpu: 1, memory: 2, disk: 50 }"
+        :medium="{ cpu: 2, memory: 4, disk: 100 }"
+      />
```
this solution shouldn't contain a vm, it should be only a gateway workload

AhmedHanafy725 commented on packages/playground/src/weblets/tf_expose.vue:
```diff
@@ -0,0 +1,242 @@
+          :rules="[validators.required('Public IP is required.'), validators.isIP('Public IP is not valid.')]"
+          #="{ props }"
+        >
+          <v-text-field label="Public IP" v-model="ip" v-bind="props" />
```
it doesn't have to be a public ip. Mycelium and Yggdarsil are not public. let's rename it to `IP`

AhmedHanafy725 commented on packages/playground/src/weblets/tf_expose.vue:
```diff
@@ -0,0 +1,242 @@
+
+    await deployGatewayName(grid, selectionDetails.value.domain, {
+      subdomain: subdomain.value,
+      ip: ip.value,
```
the ipv4 here will work as it will be in the format `http://{ip}:{port}`
but in case of the ipv6, it won't work as it should be `http://[{ip}]:{port}`

maayarosama committed:
Merge branch 'development' into development_expose_application

maayarosama committed:
Removing vm deployment

maayarosama committed:
Makint network optional in gateway

maayarosama committed:
Changing expose to domains and fixing open dialouge feature

maayarosama committed:
Merge branch 'development' into development_expose_application

MohamedElmdary committed:
feat: Add support for listing domains(gateways)
feat: Add support in deployment list to handle domains

maayarosama committed:
Fixing deployment dialouge

0oM4R reviewed (commented):
I can't delete a created domain instance,

[Screencast from 2024-08-19 10-58-34.webm](https://github.com/user-attachments/assets/149fd163-153d-47a9-9577-baada4103926)

0oM4R commented:
I tried to use custom domain, but i found this after deployment succeed

![Screenshot from 2024-08-19 11-15-31](https://github.com/user-attachments/assets/b97f1a68-eeff-4805-8e4e-d981826f0ff6)


![Screenshot from 2024-08-19 11-15-43](https://github.com/user-attachments/assets/b36a7f79-c410-4935-9166-ec6546df361e)

deployment data:
![Screenshot from 2024-08-19 11-17-06](https://github.com/user-attachments/assets/37c92193-1ad4-4393-a992-0064fe9f8b68)

0oM4R commented:
![image](https://github.com/user-attachments/assets/71794ea1-64b0-4519-add1-e08b85c38607)

I suggest to use `Domain` not `FQDN`

amiraabouhadid commented:
is localhost ip supposed to pass validation? 
![image](https://github.com/user-attachments/assets/cf0c0c25-47d9-4665-a873-a46170b2175a)

amiraabouhadid reviewed (changes requested):
tried to deploy, it just got stuck at this 
![image](https://github.com/user-attachments/assets/175bb412-48ff-4428-9dff-1af4af1ce6a2)

maayarosama committed:
Fixing delete gateways and rollbackgateway

MohamedElmdary committed:
fix: deployment data dialog

AhmedHanafy725 commented on packages/playground/public/info/domains.md:
```diff
@@ -0,0 +1,5 @@
+title: Domains
+---
+
+Domains allows users to securely expose servers hosted on local machines or VMs to the public internet. Users are required to specify the machine's IP, which can be a Mycelium IP, Yggdrasil IP, or a public IP (IPv4 or IPv6). For more details, check the [Domains documentation](https://www.manual.grid.tf/documentation/dashboard/solutions/expose.html).
```
please update the documentation file name in `info_grid` repo.

AhmedHanafy725 commented on packages/playground/src/constants/deployment_list.ts:
do we need the changes in this file?

AhmedHanafy725 commented on packages/playground/src/utils/delete_deployment.ts:
```diff
@@ -68,6 +68,9 @@ export async function deleteGateway(grid: GridClient, name: string) {
   const { deleted } = await grid.gateway.delete_name({ name });
   return deleted.length > 0;
 }
+export async function deleteGatewayContract(grid: GridClient, name: string) {
```
this method deletes a gateway deployment. it deletes the name and node contract for this gateway deployment.
```suggestion
export async function deleteGatewayDeployment(grid: GridClient, name: string) {
```

AhmedHanafy725 commented on packages/playground/src/weblets/tf_domains.vue:
```diff
@@ -0,0 +1,130 @@
+            validators.isAlphanumeric('Subdomain should consist of letters and numbers only.'),
+            subdomain => validators.isAlpha('Subdomain must start with alphabet char.')(subdomain[0]),
+            validators.minLength('Subdomain must be at least 4 characters.', 4),
+            subdomain => validators.maxLength('Subdomain cannot exceed 50 characters.', 50)(subdomain),
```
can we add async rule to check the subdomain availability like here? https://github.com/threefoldtech/tfgrid-sdk-ts/blob/03ea91378fa3b74ceb5a60a0c6228e14f9559196/packages/playground/src/components/manage_gateway_dialog.vue#L96

AhmedHanafy725 commented:
please fix the failing build

maayarosama committed:
Fix build error, adding validate subdomain and renaming deleteGatewayContract method

0oM4R commented:
![image](https://github.com/user-attachments/assets/5fb70dcf-4a00-435f-97f5-a507fd60881f)


still exists

0oM4R commented:
![image](https://github.com/user-attachments/assets/7f4e6a4a-d0f2-449a-af55-85e67c9e7636)



> I tried to use custom domain, but i found this after deployment succeed
> 
> ![Screenshot from 2024-08-19 11-15-31](https://private-user-images.githubusercontent.com/62248851/359031180-b97f1a68-eeff-4805-8e4e-d981826f0ff6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNDAzMDgsIm5iZiI6MTcyNDE0MDAwOCwicGF0aCI6Ii82MjI0ODg1MS8zNTkwMzExODAtYjk3ZjFhNjgtZWVmZi00ODA1LThlNGUtZDk4MTgyNmYwZmY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIwVDA3NDY0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkOTk5NDg3MGRjZGNlNzU3NThhMTg0OGIwMzBkNzFjNWNjZWFlOGExNTIzMjlkNWY5ZTk5ZmMyMjIzYjQ0ZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zBUVwFoq1d2uUwqYUh0i17YTHTcBn9jBE4HZ8tKFRks)
> 
> ![Screenshot from 2024-08-19 11-15-43](https://private-user-images.githubusercontent.com/62248851/359031194-b36a7f79-c410-4935-9166-ec6546df361e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNDAzMDgsIm5iZiI6MTcyNDE0MDAwOCwicGF0aCI6Ii82MjI0ODg1MS8zNTkwMzExOTQtYjM2YTdmNzktYzQxMC00OTM1LTkxNjYtZWM2NTQ2ZGYzNjFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIwVDA3NDY0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyZWI0ZTdhNzk1Mjg2OTYxYzgxODBmOGM0YTZkY2YwN2Q4ZDFiYzVkZGZhZDdhYTkyYTkyNWM1NWNkMWY4YzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MVDAOg7S4FucOVjNV4bgL3O4RH9JQHGZ2VEy_9gmA2c)
> 
> deployment data: ![Screenshot from 2024-08-19 11-17-06](https://private-user-images.githubusercontent.com/62248851/359031202-37c92193-1ad4-4393-a992-0064fe9f8b68.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNDAzMDgsIm5iZiI6MTcyNDE0MDAwOCwicGF0aCI6Ii82MjI0ODg1MS8zNTkwMzEyMDItMzdjOTIxOTMtMWFkNC00MzkzLWE5OTItMDA2NGZlOWY4YjY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIwVDA3NDY0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc5MjE5ZDVkY2JmMmU3MWJiZmFmMWQ5MDE0MzAyM2YwNGYwNTNmNmM2ZjhlYTY2YmFjMmQ3MjYzOTIxMDBjZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kktxf42fAYZ2chVlZfxfLG3sgij4Rt2U3NWUX883WPE)

MohamedElmdary committed:
- style: Capitalize state/health of gateway
- chore: Update default value for 'documentScrollend' to 'false' as 'true' casuing it to not working fine in most cases
- fix: Domains solution now show data after deploying

0oM4R commented:
> I tried to use custom domain, but i found this after deployment succeed
> 
> ![Screenshot from 2024-08-19 11-15-31](https://private-user-images.githubusercontent.com/62248851/359031180-b97f1a68-eeff-4805-8e4e-d981826f0ff6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNDgzNDYsIm5iZiI6MTcyNDE0ODA0NiwicGF0aCI6Ii82MjI0ODg1MS8zNTkwMzExODAtYjk3ZjFhNjgtZWVmZi00ODA1LThlNGUtZDk4MTgyNmYwZmY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIwVDEwMDA0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY0NWU5Zjg2MjJjMmY1YTBjZTAyZTNmNDIyYTZiYWFjYzkxNDI0NmMxOWQ3MDI4MWNiNzA3ZWMxMjczNTY0ZGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9u0zIMKtVoyOGz0E3_X4LO3_LEP1lIsWHimrnIl2eFc)
> 
> ![Screenshot from 2024-08-19 11-15-43](https://private-user-images.githubusercontent.com/62248851/359031194-b36a7f79-c410-4935-9166-ec6546df361e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNDgzNDYsIm5iZiI6MTcyNDE0ODA0NiwicGF0aCI6Ii82MjI0ODg1MS8zNTkwMzExOTQtYjM2YTdmNzktYzQxMC00OTM1LTkxNjYtZWM2NTQ2ZGYzNjFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIwVDEwMDA0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJkYTJmNzA3YTJiZDIyNmMxODFiYjY0OTNlMTQ1NjVmZDFjMjQxMWVmZWE1M2Y0OTdjYjQzMTEyYzdkZDJjMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mI2ORzfW1A7--3rCv40ESZDjtXp-PaYwOAmMmmvEll0)
> 
> deployment data: ![Screenshot from 2024-08-19 11-17-06](https://private-user-images.githubusercontent.com/62248851/359031202-37c92193-1ad4-4393-a992-0064fe9f8b68.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNDgzNDYsIm5iZiI6MTcyNDE0ODA0NiwicGF0aCI6Ii82MjI0ODg1MS8zNTkwMzEyMDItMzdjOTIxOTMtMWFkNC00MzkzLWE5OTItMDA2NGZlOWY4YjY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIwVDEwMDA0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFjZTFiZWMxMWJkM2E0ZGY0ZWQwODhjZjJmNmI2YTkyOTdjNDIzNjMzODY0ZjkwZTQ0NzBhNmUxN2ZkM2VhMDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cwKwjb8gLdzaQ4qT_6LCZG4qUZqK_1RP_Gxec4TwP-g)

[Screencast from 2024-08-20 13-06-55.webm](https://github.com/user-attachments/assets/73d765cb-e6e6-4cbe-a311-57c21666b834)

0oM4R reviewed (changes requested):
the behavior of deleting the deployments is not consist with other solutions.

[Screencast from 2024-08-20 13-12-44.webm](https://github.com/user-attachments/assets/46256cc6-ff1e-43fd-8568-3207566f4489)

mik-tf commented:
Some text suggestions:

![image](https://github.com/user-attachments/assets/4a880aab-337b-434e-8bc6-d486e7485136)

- Alternative: not points but point

![image](https://github.com/user-attachments/assets/9eef61d7-04fd-4df5-bf56-78c864faf696)

- This should be two sentences (missing a dot). Proposition: "Machine's IP: It could be a Mycelium IP, an Yggdrasil IP or a public IP (IPV4 or IPv6)"

General notes: some of those details button have the texts ending with a dot and others don't. We might want to generalize the presenting (ending or not with a dot for all).

![image](https://github.com/user-attachments/assets/27ee46db-8dc6-45fd-9cf9-9ebbed857e4a)


![image](https://github.com/user-attachments/assets/5c979f22-b29a-405c-92aa-e450aa89425e)

- Proposition for below: "The port used..."
![image](https://github.com/user-attachments/assets/48e66fdc-55f2-42d7-98d7-66c8c6f6ac30)

- "Selecting Custom Domain sets the subdomain as the gateway name"

![image](https://github.com/user-attachments/assets/a6aa097e-e12e-4583-b0dc-74255b1abe5b)

MohamedElmdary committed:
chore: applying typos suggestions

MohamedElmdary committed:
fix: Data wasn't correctly picked when using custom domain

0oM4R commented:
![image](https://github.com/user-attachments/assets/d3774919-8a36-4f58-85c0-66b3bcb0a95a)


why do we need the sshkey field here is it used ?

0oM4R commented:
please add tls passthrough option field

MohamedElmdary committed:
- feat: Add support for tls passthrough
- chore: removed unused ssh

0oM4R reviewed (approved):
Thank you both for the great work

0oM4R commented on packages/playground/src/constants/deployment_list.ts:
```diff
@@ -186,4 +191,8 @@ export const deploymentListEnvironments = {
     NODE_PILOT_HOSTNAME: "Node Pilot Hostname",
   },
+  domains: {
+    SSH_KEY: _ssh,
```
please remove this, i think we don't need it

MohamedElmdary committed:
chore: Remove unused env variables

MohamedElmdary committed:
fix: build failed due to removing env which being used somewhere in code

amiraabouhadid reviewed (commented):
why can't i toggle the TLS switch ? 
https://www.loom.com/share/8f66f5c0898246dfab389c4b7a2687b4?sid=80b3e425-10a8-4aa6-8523-093e98f71d07

amiraabouhadid commented:
> why can't i toggle the TLS switch ? https://www.loom.com/share/8f66f5c0898246dfab389c4b7a2687b4?sid=80b3e425-10a8-4aa6-8523-093e98f71d07

my bad didn't realize this switch is read only
